File: SpringCommerce/src/main/java/com/example/demo/Service/Implement/VoucherService.java
package com.example.demo.Service.Implement;

import com.example.demo.Model.*;
import com.example.demo.Repository.*;
import com.example.demo.Service.InterfaceOrderService;
import com.example.demo.nntFolder.Observer;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class VoucherService extends BaseService<Voucher, VoucherRepository>{
    @Autowired
    UserRepository userRepository;
    private final List<Observer> observers = new ArrayList<>();

    public boolean create(Voucher voucher) throws NoSuchElementException {
        repository.save(voucher);
        return true;
    }

    public boolean updateVoucher(UUID voucherId, Voucher newVoucher) throws NoSuchElementException {
        Optional<Voucher> voucherOptional = repository.findById(voucherId);
        if (voucherOptional.isPresent()) {
            newVoucher.setId(voucherId);
            repository.save(newVoucher);
            return true;
        } else {
            throw new NoSuchElementException("voucherId: " + voucherId + "is not exists");
        }

    }


    public Voucher applyVoucher(String code, UUID userId) throws NoSuchElementException {
        Optional<Voucher> voucherOptional = repository.findByCode(code);
        boolean isUsed = repository.existsByCodeAndUsers_Id(code, userId);
        if(voucherOptional.isPresent() && !isUsed) {
            // notify observer
            System.out.println("OBSERVER: " + observers.size());
            for (Observer observer : observers) {
                System.out.println("OBSERVER: " + observer.getClass().getName());
                observer.onVoucherApply(userId, code);
            }
            observers.clear();
            return voucherOptional.get();
        } else {
            return null;
        }
    }

    public boolean addUser(UUID userId, String code) throws NoSuchElementException {
        Optional<Voucher> voucherOptional = repository.findByCode(code);
        if(voucherOptional.isPresent()) {
            Voucher voucher = voucherOptional.get();
            Optional<User> userOptional = userRepository.findById(userId);
            if(userOptional.isPresent()) {
                User user = userOptional.get();
                voucher.addUser(user);
                repository.save(voucher);
                return true;
            } else {
                throw new NoSuchElementException("userId: " + userId + " is not exists");
            }

        } else {
            throw new NoSuchElementException("voucher code: " + code + " is not exists");
        }
    }
    public void attachObserver(Observer observer) {
        observers.add(observer);

    }

    public void detachObserver(Observer observer) {
        observers.remove(observer);
    }

}

File: guitar/app/src/main/java/com/example/myapplication/api/service/VoucherService.kt
package com.example.myapplication.api.service

import com.example.myapplication.api.response.ApiResponse
import com.example.myapplication.api.response.PaginatedResponse
import com.example.myapplication.model.Voucher
import retrofit2.Call
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query
import java.util.UUID

interface VoucherService {

    @GET("/api/vouchers/page")
    fun getPage(
        @Query("pageIndex") pageIndex: Int = 0,
        @Query("pageSize") pageSize: Int = 10
    ): Call<ApiResponse<PaginatedResponse<Voucher>>>

    @GET("/api/voucher/applyVoucher")
    fun applyVoucher(
        @Query("code") code: String,
        @Query("userId") userId: UUID
    ): Call<ApiResponse<Voucher>>

    @POST("/api/voucher/insert")
    fun insert(@Body voucher: Voucher): Call<ApiResponse<Boolean>>

    @GET("/api/voucher/addUser")
    fun addUser(
        @Query("code") code: String,
        @Query("userId") userId: UUID
    ): Call<ApiResponse<Boolean>>

    @GET("/api/voucher/{id}")
    fun getById(
        @Path("id") id: UUID
    ): Call<ApiResponse<Voucher>>

    @PUT("/api/voucher/update/{id}")
    fun update(
        @Path("id") userId: UUID,
        @Body voucher: Voucher
    ): Call<ApiResponse<Boolean>>
}
